### PR TITLE
migrate data from PoolCandidate to User

### LINKED ISSUE #3148: Migrate PoolCandidate data to User model
## Description
Write a migration which copies data from existing PoolCandidates to the corresponding User object.

Copy the following fields:
- isWoman
- hasDisability
- isIndigenous
- isVisibleMinority
- hasDiploma
- languageAbility
- locationPreferences
- acceptedOperationalRequirements
- expectedSalary
- expectedClassifications
- cmoAssets

Note that some of these may require creating new rows in pivot tables.

## Acceptance Criteria
- [ ] After running this migration, any data in existing PoolCandidates which has a corresponding field in the User model is copied to the candidate's User model.

### PR DESCRIPTION
**Pull request by vd1992**
closes #3148 
not a lot of lines, just tricky logic and plenty of room for error! 

### TESTING
       STAGE ONE - SETUP
 - Checkout + run all migrations/seeder to ensure things are on the right track
 - run `php artisan migrate:rollback --step=1`
 - ensure the rolled back migration is the poolcandidate_data one

       STAGE TWO - DATA PREPARATION 
- go to classification_pool_candidate table
- pick a random entry or two, write down the pool candidate id and classification id, as well, write down the user id attached to that candidate
- next, go to users table
- pick a user, note who you selected, edit said user and set various fields to null (specifically field noted in the migration)

       STAGE THREE - DATA CLEARANCE
- go to classification_users, select all entries and delete them
- go to cmo_assets_users, select all entries and delete them

      STAGE FOUR - RUN THE MIGRATION
- run `php artisan migrate --path=/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php`

      STAGE FIVE - EVALUATING THE MIGRATION
- go to users table, go to the user you noted, ensure the fields you nulled have values 
- go classifications_users and cmo_assets_users tables, ensure they have values again
- go to classifications_users, using what you noted down, ensure the pool candidate with the user id attached to some classification was added in and remains the same

      STAGE SIX - CONGRATULATIONS FOR FOLLOWING INSTRUCTIONS

### CODE REVIEW

**Comment by vd1992** [comment]
### DEPLOYMENT
BIG MIGRATIONS

**Review by petertgiles — CHANGES_REQUESTED** [review]
It works!  A comment below to make it safer though.

**Comment by petertgiles** [comment]
Here's a query I wrote for a more thorough check.  Note that if you switch to combining ID arrays instead of overwriting it might cause a false positive.
<details>
  <summary>Click to see SQL</summary>

```
select
  case
    when u.is_woman <> pc.is_woman then 'is_woman'
    when u.has_disability <> pc.has_disability then 'has_disability'
    when u.is_indigenous <> pc.is_indigenous then 'is_indigenous'
    when u.is_visible_minority <> pc.is_visible_minority then 'is_visible_minority'
    when u.has_diploma <> pc.has_diploma then 'has_diploma'
    when u.language_ability <> pc.language_ability then 'language_ability'
    when u.location_preferences <> pc.location_preferences then 'location_preferences'
    when u.accepted_operational_requirements <> pc.accepted_operational_requirements then 'accepted_operational_requirements'
    when u.expected_salary <> pc.expected_salary then 'expected_salary'
    when cu.classification_ids <> cpc.classification_ids then 'classification_ids'
    when cau.cmo_asset_ids <> capc.cmo_asset_ids then 'cmo_asset_ids'
    else null
  end matching_error,
  u.id user_id,
  pc.id pool_candidate_id,
  u.is_woman,
  pc.is_woman,
  u.has_disability,
  pc.has_disability,
  u.is_indigenous,
  pc.is_indigenous,
  u.is_visible_minority,
  pc.is_visible_minority,
  u.has_diploma,
  pc.has_diploma,
  u.language_ability,
  pc.language_ability,
  u.location_preferences,
  pc.location_preferences,
  u.accepted_operational_requirements,
  pc.accepted_operational_requirements,
  u.expected_salary,
  pc.expected_salary,
  cu.classification_ids,
  cpc.classification_ids,
  cau.cmo_asset_ids,
  capc.cmo_asset_ids
from users u
join pool_candidates pc  on u.id = pc.user_id
left join (
  select user_id, string_agg(cast(classification_id as varchar), ', ' order by classification_id) classification_ids
  from classification_user
  group by user_id
) cu on u.id = cu.user_id
left join (
  select pool_candidate_id, string_agg(cast(classification_id as varchar), ', ' order by classification_id) classification_ids
  from classification_pool_candidate
  group by pool_candidate_id
) cpc on pc.id = cpc.pool_candidate_id
left join (
  select user_id, string_agg(cast(cmo_asset_id as varchar), ', ' order by cmo_asset_id) cmo_asset_ids
  from cmo_asset_user
  group by user_id
) cau on u.id = cau.user_id
left join (
  select pool_candidate_id, string_agg(cast(cmo_asset_id as varchar), ', ' order by cmo_asset_id) cmo_asset_ids
  from cmo_asset_pool_candidate
  group by pool_candidate_id
) capc on pc.id = capc.pool_candidate_id
```
</details>

File: api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php
Comment (vd1992) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php:39]: image

makes this happen, so no I guess?

image

Comment (vd1992) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php:39]: you may have to handhold me here if this isn't what you meant
okay, I took a stab at what I think you meant

image

Comment (petertgiles) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php:39]: Yeah, that looks right to me!

Comment (petertgiles) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php:39]: I'm still learning my Eloquent but I think this works.
image
We don't have PHP linting yet but watch the tabs.

Comment (vd1992) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php]: added in the latest commit
guess we aren't merging this yet so just let me know if everything works fine and we can merge it down the road
don't approve it either please, so it isn't accidentally merged 🤣

Comment (vd1992) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php]: how do you get those dots and things to show? I don't have it in my editor

Comment (petertgiles) [reply api/database/migrations/2022_08_02_193722_poolcandidate_data_to_user.php]: I've got an extension called "Render Line Endings" installed.

**Comment by petertgiles** [comment]
I approve!  I'll hold-off "github approving" until we're ready to merge this.

**Comment by vd1992** [comment]
merging waiting until Applicant Search live

**Pull request merged**
